keps/sig-api-machinery/5589-gogo-dependency/README.md

```diff
@@ -0,0 +1,640 @@
+* v1.7: make map serialization deterministic ([#47701](https://github.com/kubernetes/kubernetes/pull/47701))
+* 2020: gogo / golang/protobuf incompatibilities stall dependency updates, panic etcd libraries, get resolved in a best-effort way
+* 2021-2022: gogo transitions to deprecated / unmaintained
+* v1.35: This KEP proposed
+
+## Drawbacks
+
+Taking on responsibility for complex protobuf generation code is not ideal.
+However, I would claim that the Kubernetes project already has implicit responsibility
+for the *output* of the unmaintained generator, so I don't think a maintenance-mode fork
+whose only consumer is the Kubernetes API protobuf generator is significantly different
+than the current state.
+
+## Alternatives
+
+Two alternatives were considered:
+
+1. Switching to standard `protoc`/`protoc-gen-go` generation.
+   The impact to non-protobuf use of Kubernetes API go types,
+   and the mismatch between Kubernetes requirements around canonical serialization
+   and dropping of unknown fields ruled this out. See the [Notes/Constraints/Caveats](#notesconstraintscaveats)
+   section for more details.
+
+2. Switching to another non-standard protobuf generation (e.g. `vtprotobuf`).
+   This would have required significantly more work to achieve,
+   and would have been significantly more difficult to verify compatibility.
+   There were not clear benefits that would justify this amount of time investment.
+   The approach described in this KEP does not preclude such a change in the future,
+   if it becomes apparent there are benefits that would justify a change of that magnitude,
+   and a plan is proposed to make that change safely.
+
+## Possible Future Work
+
+The following are *possible* future changes that could be made to the protobuf generator
+once it is in a location that can be modified in ways scoped to Kubernetes API types.
+These changes are not planned as part of this KEP, but are captured here for reference.
+
+**Add omitzero support**
+
+Current protobuf serialization *always* outputs non-pointer scalar and struct fields.
+This means that feature-gate-disabled fields must be created as pointers so they will not be serialized unless populated.
+Adding support for an `omitzero` tag to protobuf serialization would allow for non-pointer fields which can still be feature-gate-disabled.
+
+**Simplify String() implementation**
+
+There is widespread reliance on Kubernetes API objects implementing `String()`,
+so preserving the existence of `String()` implementations is necessary to minimize disruption.
+
+However, this only *happens* to be done as part of protobuf generation,
+and the output is a pseudo-Go structure that requires large amounts of code to construct,
+and is not particularly meaningful from an API perspective.
+
+After the code-generation is forked and modifiable, the `String()` implementation could be
+switched to a simpler implementation, for example returning a JSON encoding of the object.
```

Comment:
The presence of those `String` implementations is making logging of those types annoying: instead of having the types serialized by the logger, we end up with getting the `String` method called and the less readable hard-coded serialization. `klog.Format`'s only purpose is to hide the `String` method, but needs to be added manually at the log call site.

One inconsistency is also that the `String` method is defined for pointers, but not for values. Logging a *value* doesn't get the method called.

It would be nice to just drop it in the generated code, but I agree that we can't.

What may work is *adding* something, for example `MarshalLog`. Our logging backend then should favor that over `String`. I look forward to this KEP being implemented to try this out... 😄